python/ray/serve/tests/test_failure.py

```diff
@@ -157,44 +155,45 @@ def __call__(self, *args):
 def test_worker_replica_failure(serve_instance):
     client = serve_instance
 
+    @ray.remote
+    class Counter:
+        def __init__(self):
+            self.count = 0
+
+        def inc_and_get(self):
+            self.count += 1
+            return self.count
+
     class Worker:
         # Assumes that two replicas are started. Will hang forever in the
         # constructor for any workers that are restarted.
-        def __init__(self, path):
+        def __init__(self, counter):
             self.should_hang = False
-            if not os.path.exists(path):
-                with open(path, "w") as f:
-                    f.write("1")
-            else:
-                with open(path, "r") as f:
-                    num = int(f.read())
-
-                with open(path, "w") as f:
-                    if num == 2:
-                        self.should_hang = True
-                    else:
-                        f.write(str(num + 1))
-
-            if self.should_hang:
+            self.index = ray.get(counter.inc_and_get.remote())
+            if self.index > 2:
                 while True:
                     pass
 
         def __call__(self, *args):
-            pass
+            return self.index
 
-    temp_path = os.path.join(tempfile.gettempdir(),
-                             serve.utils.get_random_letters())
-    client.create_backend("replica_failure", Worker, temp_path)
+    counter = Counter.remote()
+    client.create_backend("replica_failure", Worker, counter)
     client.update_backend_config(
         "replica_failure", BackendConfig(num_replicas=2))
     client.create_endpoint(
         "replica_failure", backend="replica_failure", route="/replica_failure")
 
     # Wait until both replicas have been started.
     responses = set()
-    while len(responses) == 1:
-        responses.add(request_with_retries("/replica_failure", timeout=1).text)
+    start = time.time()
+    while time.time() - start < 30:
         time.sleep(0.1)
+        responses.add(request_with_retries("/replica_failure", timeout=1).text)
+        if len(responses) > 1:
+            break
+    else:
+        raise TimeoutError("Timed out waiting for replicas after 30s.")
 
```

Comment:
@edoakes Should we actually check that we get some response?

Reply:
good call, let me add an assert